example-async.js

```diff
@@ -0,0 +1,149 @@
+'use strict'
+
+const Fastify = require('fastify')
+
+function build (opts) {
+  const fastify = Fastify(opts)
+
+  fastify
+    .register(require('fastify-jwt'), { secret: 'supersecret' })
+    .register(require('fastify-leveldb'), { name: 'authdb-async' })
+    .register(require('./fastify-auth'))
+    .after(routes)
+
+  fastify.decorate('verifyJWTandLevel', verifyJWTandLevel)
+  fastify.decorate('verifyUserAndPassword', verifyUserAndPassword)
+
+  function verifyJWTandLevel (request, reply) {
+    const jwt = this.jwt
+    const level = this.level
+
+    if (request.body && request.body.failureWithReply) {
+      reply.code(401).send({ error: 'Unauthorized' })
+      return Promise.reject(new Error())
+    }
+
+    if (!request.req.headers['auth']) {
+      return Promise.reject(new Error('Missing token header'))
+    }
+
+    return new Promise(function (resolve, reject) {
+      jwt.verify(request.req.headers['auth'], function (err, decoded) {
+        if (err) { reject(err) };
+        resolve(decoded)
+      })
+    }).then(function (decoded) {
+      return level.get(decoded.user)
+        .then(function (password) {
+          if (!password || password !== decoded.password) {
+            throw new Error('Token not valid')
+          }
+        })
+    }).catch(function () {
+      throw new Error('Token not valid')
```

Comment:
(missing the err parameter)
```suggestion
      throw new Error(err.message)
```

Reply:
I think using “Token not valid” is a security feature and it should be retained

Reply:
Of course, but the error is thrown in a "then" so we don't lost this behavior, we only add a check that there are not other masquerade errors. 
This was my intent here

Reply:
We can log the error and rethrow. The errors are purposefully masked so that they don't baloon in the Fastify error handler and then displayed to the user.

Reply:
Current state is kind of a successor of this approach: https://github.com/fastify/fastify-auth/blob/master/example.js#L29 , so there is not real error but generic one.

Added the log before the error is thrown.